### ISSUE TITLE
Allow non-tiled aligned shapes for LayerNorm

### ISSUE THREAD

**Issue by bbradelTT** [tenstorrent/tt-metal#28365]
See https://github.com/tenstorrent/tt-metal/issues/18548 and all of that issue's comments for details.

It was found that LayerNorm only handles widths that are tile aligned. Needs to handle all shapes.

Repro steps:

Run the following pytest.
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten

def test_layer_norm(device):
    dtype = ttnn.bfloat16
    dlayout = ttnn.TILE_LAYOUT
    epsilon = 1e-6

    # fail 196, 38
    # pass 32, 16, 20
    h = 1
    # fail 768
    # pass 32, 38
    w = 2 
    x = torch.zeros(1, h, w).to(torch.bfloat16)
    for i in range(h):
        for j in range(w):
            if (i + j) % 2 == 0:
                x[0,i,j] = 1
    weight = torch.ones([w]).to(torch.bfloat16)
    bias = torch.zeros([w]).to(torch.bfloat16)
    try:
        print(f'DATA TYPE {dtype} SHAPE {x.shape}')
        ref_value = aten.native_layer_norm.default(x, [w], weight, bias, epsilon)[0]
        
        tt_x = ttnn.from_torch(x, dtype=dtype, layout=dlayout, device=device)
        tt_weight = ttnn.from_torch(weight, dtype=dtype, layout=dlayout, device=device)
        tt_bias = ttnn.from_torch(bias, dtype=dtype, layout=dlayout, device=device)

        tt_result = ttnn.layer_norm(tt_x, epsilon=epsilon)
        tt_result = ttnn.to_torch(tt_result)
        
        inds = torch.where(ref_value != tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    print(f"ref_value {ref_value}\n tt_result {tt_result}")
    assert_with_pcc(ref_value, tt_result, 0.999)
```

Expected Output:
Same as torch: -1, 1

Actual:
Something different: 5.5625, -0.1797

Proof of concept to get the correct results:

1. The 1 / W (padded) needs to be changed to 1 / non padded W
```
diff --git a/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp b/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp
index b4488a65f5..954e0587a7 100644
--- a/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp
+++ b/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp
@@ -449,7 +449,7 @@ operation::ProgramWithCallbacks layernorm_multi_core(
     }
 
     uint32_t curr_row = 0;
-    float winv = 1.0f / W;
+    float winv = 1.0f / a.logical_shape()[-1];
```

2. The reduce / matmul scaler needs to be limited to an arbitrary limit.
For matmul that means the code needs to be changed in the following way.
```
ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_mm_scaler.hpp:FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler) {

FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler) {

    for (int i = 0; i < 128; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }
    for (int i = 256; i < 384; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }

FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler, uint32_t n) {

    int face0_bound = (n < 16)? (n * 8) : 128;
    int face2_bound = (n < 16)? 256 : (256 + (n-16) * 8);

    for (int i = 0; i < face0_bound; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }       
    for (int i = 256; i < face2_bound; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }      
```
The diff is
```
--- a/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_mm_scaler.hpp
+++ b/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_mm_scaler.hpp
@@ -35,3 +35,37 @@ FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler
 
     cb_push_back(cb_id, 1);
 }
+
+
+// Tile is assumed to have 16-bit elements
+// Scaler is assumed to be a 16-bit value double packed into a u32
+FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler, const uint32_t n) {
+    cb_reserve_back(cb_id, 1);
+
+    constexpr uint32_t num_zeros_reads = 2048 / MEM_ZEROS_SIZE;
+    uint64_t zeros_noc_addr = get_noc_addr(MEM_ZEROS_BASE);
+    uint32_t write_addr = get_write_ptr(cb_id);
+    volatile tt_l1_ptr uint32_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(write_addr);
+
+    // Fill tile with zeros
+    // TODO: src addr does not need to be rewritten. Update/add api for this
+    noc_async_read_one_packet_set_state(zeros_noc_addr, MEM_ZEROS_SIZE);
+    for (uint32_t i = 0; i < num_zeros_reads; ++i) {
+        noc_async_read_one_packet_with_state(zeros_noc_addr, write_addr);
+        write_addr += MEM_ZEROS_SIZE;
+    }
+    noc_async_read_barrier();
+
+    int face0_bound = (n < 16)? (n * 8) : 128;
+    int face2_bound = (n < 16)? 256 : (256 + (n-16) * 8);
+
+    uint32_t single_packed_scalar = scaler & 0xFFFF;
+    for (int i = 0; i < face0_bound; i += 8) {
+        ptr[i] = single_packed_scalar;
+    }
+    for (int i = 256; i < face2_bound; i += 8) {
+        ptr[i] = single_packed_scalar;
+    }
+
+    cb_push_back(cb_id, 1);
+}
```
For reduce, the code needs to be
```
        int outer_bound = (n < 16)? 1 : 2;
        int inner_bound = (n < 16)? n/2 : 8;

        for (int k = 0; k < outer_bound; ++k) {
            uint32_t idx = k << 7; // * 128
            for (int j = 0; j < inner_bound; ++j) {
                ptr[idx + j] = scaler;
            }
            inner_bound = (n - 16)/2;
        }
```
The diff is
```
--- a/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_reduce_scaler.hpp
+++ b/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_reduce_scaler.hpp
@@ -36,6 +36,38 @@ FORCE_INLINE void generate_reduce_scaler(const uint32_t cb_id, const uint32_t sc
 
     cb_push_back(cb_id, 1);
 }
+FORCE_INLINE void generate_reduce_scaler(const uint32_t cb_id, const uint32_t scaler, const uint32_t n) {
+    cb_reserve_back(cb_id, 1);
+
+    constexpr uint32_t num_zeros_reads = 2048 / MEM_ZEROS_SIZE;
+    uint64_t zeros_noc_addr = get_noc_addr(MEM_ZEROS_BASE);
+    uint32_t write_addr = get_write_ptr(cb_id);
+    volatile tt_l1_ptr uint32_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(write_addr);
+
+    // Fill tile with zeros
+    // TODO: src addr does not need to be rewritten. Update/add api for this
+    noc_async_read_one_packet_set_state(zeros_noc_addr, MEM_ZEROS_SIZE);
+    for (uint32_t i = 0; i < num_zeros_reads; ++i) {
+        noc_async_read_one_packet_with_state(zeros_noc_addr, write_addr);
+        write_addr += MEM_ZEROS_SIZE;
+    }
+    noc_async_read_barrier();
+
+    if (scaler != 0) {
+        int outer_bound = (n < 16)? 1 : 2;
+        int inner_bound = (n < 16)? n/2 : 8;
+
+        for (int k = 0; k < outer_bound; ++k) {
+            uint32_t idx = k << 7; // * 128
+            for (int j = 0; j < inner_bound; ++j) {
+                ptr[idx + j] = scaler;
+            }
+            inner_bound = (n - 16)/2;
+        }
+    }
+
+    cb_push_back(cb_id, 1);
+}
```
I just hard coded the limits to 2

ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/dataflow/reader_unary_interleaved_ln.cpp
```
        generate_mm_scaler(cb_in_2, scaler, 2);
#else
        generate_reduce_scaler(cb_in_2, scaler, 2);
```

This issue will focus on the reduce only. A separate issue will need to be created to look at matmul as the last task of this issue.
